### PR TITLE
Add the Sawmill

### PR DESCRIPTION
**Pull request by BluSunrize**
It stripes logs, then cuts them into planks.
I think I covered everything in there, changelog comes after merge.

Still missing a subtitle for the manual, but that can come later. Also hopefully getting a better sawdust texture from Hazard.

I did find a bug with IngredientFluidStack and shapeless recipes, which is fixed in:
https://github.com/BluSunrize/ImmersiveEngineering/commit/22731efa28d131df7765e70bdab32dd35cd199da
We need to monitor if that fix breaks on 1.16+ due to the timing at which tags load.

### CODE REVIEW

**Comment by malte0811** [comment]
> I did find a bug with IngredientFluidStack and shapeless recipes, which is fixed in: 22731ef

Can you elaborate on that? I'm pretty sure we need to override `isSimple` as "simple" means "only cares about the item ID". The `hasNoMatchingItems` looks reasonable, though I'd expect the default implementation to be something like `getMatchingStacks().size() == 0`, which should work here.

File: src/main/java/blusunrize/immersiveengineering/client/render/tile/SawmillRenderer.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	@Override
+	public void render(SawmillTileEntity te, float partialTicks, MatrixStack matrixStack, IRenderTypeBuffer bufferIn, int combinedLightIn, int combinedOverlayIn)
+	{
+		if(!te.formed||te.isDummy()||!te.getWorldNonnull().isBlockLoaded(te.getPos()))
+			return;
+
+		//Grab model
+		final BlockRendererDispatcher blockRenderer = Minecraft.getInstance().getBlockRendererDispatcher();
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/client/render/tile/SawmillRenderer.java]: Unused variable?

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+		super.tick();
+		if(isDummy()||isRSDisabled())
+			return;
+		if(world.isRemote&&shouldRenderAsActive())
+		{
+			animation_bladeRotation += 36f;
+			animation_bladeRotation %= 360f;
+		}
+		if(world.isRemote)
+			return;
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java:151]: I would prefer something like
```java
if (world.isRemove)
{
   if (shouldRenderAsActive())
   {…}
   return;
}
```
but this is fine too.

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+		for(ItemStack output : secondaries)
+			doSecondaryOutput(output.copy());
+	}
+
+	@Override
+	public boolean interact(Direction side, PlayerEntity player, Hand hand, ItemStack heldItem, float hitX, float hitY, float hitZ)
+	{
+		SawmillTileEntity master = master();
+		if(master!=null)
+			if(player.isSneaking()&&!master.sawblade.isEmpty())
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java:187]: I've said this before, I don't really like `if(a) if(b) … else` without braces on the outer if. Could you add those?

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+					player.setHeldItem(hand, master.sawblade.copy());
+				else if(!world.isRemote)
+					player.entityDropItem(master.sawblade.copy(), 0);
+				master.sawblade = ItemStack.EMPTY;
+				this.updateMasterBlock(null, true);
+				return true;
+			}
+			else if(IETags.sawblades.contains(heldItem.getItem()))
+			{
+				ItemStack tempMold = !master.sawblade.isEmpty()?master.sawblade.copy(): ItemStack.EMPTY;
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java]: I would guess you copied this from the metal press and forgot to rename it?

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+				master.sawblade = Utils.copyStackWithAmount(heldItem, 1);
+				heldItem.shrink(1);
+				if(heldItem.getCount() <= 0)
+					heldItem = ItemStack.EMPTY;
+				else
+					player.setHeldItem(hand, heldItem);
+				if(!tempMold.isEmpty())
+					if(heldItem.isEmpty())
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java:208]: Same comment on `if(a)if(b) else` as before.

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+			SawmillTileEntity master = master();
+			if(master==null)
+				return;
+			ItemStack stack = ((ItemEntity)entity).getItem();
+			if(stack.isEmpty())
+				return;
+			master.insertItemToProcess(stack, false);
+			if(stack.getCount() <= 0)
+				entity.remove();
+		}
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java:384]: I'm disappointed, stepping on the giant exposed spinning sawblade does not hurt you?

**Review by malte0811 — CHANGES_REQUESTED** [review]

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java (hunk 6 above)
Comment (BluSunrize) [reply src/main/java/blusunrize/immersiveengineering/common/blocks/metal/SawmillTileEntity.java:384]: ...you make a good point.

**Comment by BluSunrize** [comment]
Ready for another review

**Review by malte0811 — APPROVED** [review]
Looks good now.

**Pull request merged**